typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,1655 @@
+This happens because of information loss.
+The type of an object with `readonly` is a subtype of the type without it, and when you pass it to a function with a more general type, the function does not have access to information about its subtype.
+
+```typescript
+const a: { readonly x: number } = { x: 1 };
+
+const changeA = (arg: { x: number }): void => {
+  const argC = arg;
+  argC.x = 2;
+}
+
+changeA(a);
+a.x; // 2
+```
+
+For this reason, you should always write proper types of arguments.
+
+
+#### `Readonly` type
+
+With the [`Readonly`](https://www.typescriptlang.org/docs/handbook/utility-types.html#readonlytype) type, you can mark all properties as `readonly`.
+
+```typescript
+type A = {
+  x: number;
+  y: number;
+};
+
+type ImmutableA = Readonly<A>; 
+const a: ImmutableA = { x: 1, y: 2 };
+a.x = 2; // Error: Cannot assign to 'x' because it is a read-only property.
+a.y = 1; // Error: Cannot assign to 'y' because it is a read-only property.
+```
+
+With `ReadonlyArray`, you can make all array elements `readonly`.
+
+```typescript
+const arr: ReadonlyArray<number> = [1, 2, 3];
+arr[0] = 4; // Error: Index signature in type 'readonly number[]' only permits reading.
+```
+
+You can also make a `readonly` dictionary, for example:
+
+```typescript
+type A = {
+    readonly [x: string]: number;
+}
+
+const a: A = { x: 1, y: 2 };
+a.x = 2; // Error: Index signature in type 'A' only permits reading.
+a.y = 1; // Error: Index signature in type 'A' only permits reading.
+```
+
+Now you have a more flexible way to make things read-only.
+
+#### Working with immutable objects
+
+Although you can now make immutable objects, usually you don't need them as they are.
+You want to do some actions on them and make new immutable objects.
+To do that, you can construct new objects by hand or use the `Writable` utility type, but it is not that simple.
+
+Fortunately, there are libraries that can help you deal with this issue.
+
+The most powerful is [immer](https://immerjs.github.io/immer/), which is used for immutable state flow.
+Immutable data there is not copied but shared in memory.
+In other words, `immer` gives you the ability to work with a draft copy of your data and not worry about mutability.
+After all the changes are done, it will produce an actually immutable state.
+It also has helpers for React.
+
+Here's an example of its use:
+
+```typescript
+type A = Immutable<
+  {
+    text: string;
+    valid: boolean;
+  }[]
+>;
+
+const a: A = [
+  {
+    text: "valid text",
+    valid: true,
+  },
+  {
+    text: "invalid text",
+    valid: false,
+  },
+];
+
+const nextA = produce(a, (draft) => {
+  draft[1].valid = false;  // set valid = false for the second element
+  draft.push({ text: "some text", valid: false }); // add new element to array
+});
+```
+
+In the code example above, `produce` is a base primitive for working with `immer`.
```

Comment:
I don't really know how it works under the hood, but what for it may need `Writable`? Search in their repo by `Writable` said that there is no `Writable`